subsys/nfc/ndef/ch_rec_parser.c

```diff
@@ -0,0 +1,585 @@
+				u8_t *result_buf, u32_t *result_buf_len)
+{
+	int err;
+	const u8_t *payload_buf = payload_desc->payload;
+	u32_t payload_len = payload_desc->payload_length;
+	const u32_t buf_size = *result_buf_len;
+	struct net_buf_simple buf;
+	struct nfc_ndef_ch_ac_rec *ac_rec;
+
+	/* Initialize memory allocator with buffer memory block and
+	 * its size.
+	 */
+	net_buf_simple_init_with_data(&buf, result_buf, buf_size);
+	net_buf_simple_reset(&buf);
+
+	ac_rec = (struct nfc_ndef_ch_ac_rec *)
+			memory_allocate(&buf, sizeof(*ac_rec));
+	if (!ac_rec) {
+		return -ENOMEM;
+	}
+
+	memset(ac_rec, 0, sizeof(*ac_rec));
+
+	if (payload_len < AC_REC_CPS_BYTE_SIZE) {
+		return -EINVAL;
+	}
+
+	ac_rec->cps = (enum nfc_ndef_ch_ac_rec_cps)(*payload_buf);
+	payload_buf += AC_REC_CPS_BYTE_SIZE;
+	payload_len -= AC_REC_CPS_BYTE_SIZE;
+
+	err = ac_reference_parse(&ac_rec->carrier_data_ref, &buf,
+				 &payload_buf, &payload_len);
+	if (err) {
+		return err;
+	}
+
+	if (payload_len < 1) {
+		return -EINVAL;
+	}
+
+	ac_rec->aux_data_ref_cnt = *payload_buf;
+	payload_buf++;
+	payload_len--;
+
+	if (!ac_rec->aux_data_ref_cnt) {
+		return 0;
+	}
+
+	ac_rec->aux_data_ref = (struct nfc_ndef_ch_ac_rec_ref *)
+					memory_allocate(&buf, ac_rec->aux_data_ref_cnt * sizeof(*ac_rec->aux_data_ref));
+	if (ac_rec->aux_data_ref) {
+		return -ENOMEM;
+	}
+
+	for (size_t i = 0; i < ac_rec->aux_data_ref_cnt; i++) {
+		err = ac_reference_parse(&ac_rec->aux_data_ref[i], &buf,
+					 &payload_buf, &payload_len);
+		if (err) {
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int ch_rec_payload_parse(struct nfc_ndef_bin_payload_desc *payload_desc,
+				u8_t *result_buf, u32_t *result_buf_len,
+				bool nlen)
+{
+	int err;
+	const u32_t buf_size = *result_buf_len;
+	const u8_t *payload_buf = payload_desc->payload;
+	u32_t payload_len = payload_desc->payload_length;
+	u32_t local_msg_size;
+	u8_t *local_msg;
+	struct nfc_ndef_ch_rec *ch_rec;
+	struct net_buf_simple buf;
+
+	/* Initialize memory allocator with buffer memory block and its size. */
+	net_buf_simple_init_with_data(&buf, result_buf, buf_size);
+	net_buf_simple_reset(&buf);
+
+	ch_rec = (struct nfc_ndef_ch_rec *)
+				memory_allocate(&buf, sizeof(*ch_rec));
+	if (!ch_rec) {
+		return -ENOMEM;
+	}
+
+	memset(ch_rec, 0, sizeof(*ch_rec));
+
+	if (payload_len < 1) {
+		return -EINVAL;
+	}
+
+	ch_rec->major_version = ((*payload_buf) >> 4) & 0x0F;
+	ch_rec->minor_version = (*payload_buf) & 0x0F;
+	payload_buf++;
+	payload_len--;
+
+	/* Skip local message NDEF Len field. */
+	if (nlen) {
+		payload_buf += TAG_TYPE_4_NLEN_FIELD_SIZE;
+		payload_len -= TAG_TYPE_4_NLEN_FIELD_SIZE;
+	}
+
+	local_msg_size = net_buf_simple_tailroom(&buf);
+
+	local_msg = memory_allocate(&buf, local_msg_size);
+	if (!local_msg) {
+		return -ENOMEM;
+	}
+
+	err = nfc_ndef_msg_parse(local_msg, &local_msg_size,
+				 payload_buf, &payload_len);
+	if (err) {
+		return err;
+	}
+
+	ch_rec->local_records = (struct nfc_ndef_msg_desc *)local_msg;
+
+	return 0;
```

Comment:
I think you forgot to update ```result_buf_len```. It should be equal to the parsed NDEF record length